### PR TITLE
feat: Evaluate missing splits

### LINKED ISSUE #1260: Only skip benchmarking if split results are the same too
We currently skip evaluation if the results file already exists, however, it may be that the file exists but not with the same splits/subsets as requested by the current evaluation. I think we may want to also check equivalence across splits & subsets and then run the ones still missing and edit the existing results file to put them back in.

### PR DESCRIPTION
**Pull request by isaac-chung**
Fixes #1260 

- Based on https://github.com/embeddings-benchmark/mteb/pull/1268 and addresses most of the comments
- Once run, the returned result will contain the missing splits as well as the already run splits
- Requires `overwrite=True` to work
- Changed tests to use the same MTEB object instead of a new one

Example Usage
```python
from mteb import MTEB, get_tasks, get_model

tasks = get_tasks(tasks=["Banking77Classification"])
model = get_model("sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2")

# Runs only one split
evaluation = MTEB(tasks=tasks)
results = evaluation.run(
    model,
    eval_splits=["val"],
    output_folder=str(tmp_path / "testcase2"),
    verbosity=2,
)

# Will run the missing split and return results of both splits.
results2 = evaluation.run(
    model,
    eval_splits=["val", "test"],
    output_folder=str(tmp_path / "testcase2"),
    verbosity=2,
    overwrite_results=True,
)
```

## Checklist

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`.

### CODE REVIEW

File: mteb/evaluation/MTEB.py
Comment (Samoed) [inline mteb/evaluation/MTEB.py:353]: Should `kg_co2_emissions` sum too?

File: mteb/evaluation/MTEB.py
Comment (Samoed) [inline mteb/evaluation/MTEB.py:349]: What we should do if existing result and new result have different version?

**Review by Samoed — COMMENTED** [review]
Overall looks great!

File: mteb/evaluation/MTEB.py
Comment (KennethEnevoldsen) [reply mteb/evaluation/MTEB.py:349]: One solution is only to extend results if the versions match.

File: mteb/evaluation/MTEB.py
Comment (KennethEnevoldsen) [inline mteb/evaluation/MTEB.py:440]: by language here you mean the subset on HF?

File: mteb/evaluation/MTEB.py
Comment (KennethEnevoldsen) [inline mteb/evaluation/MTEB.py:497]: (not really for this PR but easy to add) codecarbon is a required dependency now so we could just move this to the top

**Review by KennethEnevoldsen — APPROVED** [review]
Looks great only a few minor things

File: mteb/evaluation/MTEB.py
Comment (Samoed) [reply mteb/evaluation/MTEB.py:497]: The target branch of this PR is `main`, and in this branch, CodeCarbon is still optional

Comment (isaac-chung) [reply mteb/evaluation/MTEB.py:497]: Ah yes, the required dependency is only in v2.0.0 for now, not `main` yet.

File: mteb/evaluation/MTEB.py
Comment (isaac-chung) [reply mteb/evaluation/MTEB.py:349]: Will add a warning if the versions do not match

File: mteb/evaluation/MTEB.py
Comment (Samoed) [reply mteb/evaluation/MTEB.py]: I think this is important, but I can work on after merge this PR

Comment (isaac-chung) [reply mteb/evaluation/MTEB.py]: Great!

File: mteb/evaluation/MTEB.py
Comment (isaac-chung) [reply mteb/evaluation/MTEB.py]: This was causing a test to fail, as this could be None, and needs to be handled. I think it's best for now to assign as None if either new or existing one is None, and only add them if both are non-None.

File: mteb/evaluation/MTEB.py
Comment (Samoed) [reply mteb/evaluation/MTEB.py:360]: Then maybe to use verion from new_results?

Comment (isaac-chung) [reply mteb/evaluation/MTEB.py:360]: Sure, will do.

Separately, I don't think this currently takes the difference in _dataset version_ into consideration. I think this is already an existing gap, where we only check whether the same model + model revision has been run, but not check for dataset version. We should probably address it here before merging. wdyt?

Comment (Samoed) [reply mteb/evaluation/MTEB.py:360]: Most datasets downloaded by revision, so I don't think we need more checks

Comment (isaac-chung) [reply mteb/evaluation/MTEB.py]: Ok, then this should be good to merge once the tests pass.

**Comment by isaac-chung** [comment]
Thanks @Samoed and @KennethEnevoldsen for reviewing, and @thivyanth for the initial iteration! Merging now.

File: mteb/evaluation/MTEB.py
Comment (KennethEnevoldsen) [reply mteb/evaluation/MTEB.py]: I think it is good to merge. It might be nice to create a "deprecation version", e.g. before 1.11.0, we can then slowly outdated old results as needed. However, this is probably something for a more general discussion, before any implementation.

**Pull request merged**
